virtwho/config.py

```diff
@@ -87,6 +87,118 @@ class NotSetSentinel(object):
     pass
 
 
+class Info(object):
+    """
+    A class containing keys with values.
+    If the kwargs passed to the init of this class do not include all
+    required_args, init will fail with a ValueError.
+
+    All required and optional kwargs will be accessible on an instance of
+    this object as attributes.
+    No required or optional kwargs can start with an '_'.
+
+    This class is used as a base for creating entities that are constituted
+    entirely by a set of named values. In the case of virt-who we subclass
+    this for all destination objects in order to attempt to pull all required
+    info for a given type of destination out of configs and to help determine
+    what unique destinations exist across all configurations.
+    """
+    required_kwargs = ()
+    optional_kwargs = ()
+
+    def __init__(self, **kwargs):
+        self.__dict__["_options"] = {}
+        attributes_to_add = []
+        attributes_to_add.extend(type(self).required_kwargs)
+        if type(self).optional_kwargs:
+            attributes_to_add.extend(type(self).optional_kwargs)
+
+        for arg in attributes_to_add:
+            try:
+                self._options[arg] = kwargs[arg]
+            except KeyError:
+                if arg in type(self).required_kwargs:
+                    raise ValueError("Missing required option: %s" % arg)
+
+    def __hash__(self):
+        to_hash = []
+        for item in self._options.values():
+            if isinstance(item, list):
+                to_hash.append(tuple(item))
+            else:
+                to_hash.append(item)
+        return hash(tuple(to_hash))
+
+    def _set_option(self, key, value):
+        self._options[key] = value
+
+    def _get_option(self, key):
```

Comment:
Is this getter and the above setter needed, or can they be replaced with their implementation?